src/Infer/ADT.hs

```diff
@@ -31,69 +31,80 @@ buildADTTypes' astPath adts (adt : xs) = do
+  Just t  -> throwError $ InferError (ADTAlreadyDefined t) NoReason
+  Nothing -> return
+    (adtname adt, TComp astPath (adtname adt) (TVar . TV <$> adtparams adt))
 
 
-resolveADTs :: FilePath -> ADTs -> [ADT] -> Infer Vars
-resolveADTs astPath tadts adts =
-  mergeVars <$> mapM (resolveADT astPath tadts) adts
+resolveADTs :: Env -> FilePath -> ADTs -> [ADT] -> Infer Vars
+resolveADTs priorEnv astPath tadts adts =
+  mergeVars <$> mapM (resolveADT priorEnv astPath tadts) adts
  where
   mergeVars []   = M.empty
   mergeVars vars = foldr1 M.union vars
 
 
-resolveADT :: FilePath -> ADTs -> ADT -> Infer Vars
-resolveADT astPath tadts ADT { adtname, adtconstructors, adtparams } =
-  foldr1 M.union
-    <$> mapM (resolveADTConstructor astPath tadts adtname adtparams)
-             adtconstructors
+resolveADT :: Env -> FilePath -> ADTs -> ADT -> Infer Vars
+resolveADT priorEnv astPath tadts adt =
+  let name   = adtname adt
+      ctors  = adtconstructors adt
+      params = adtparams adt
+  in  foldr1 M.union
+        <$> mapM (resolveADTConstructor priorEnv astPath tadts name params)
+                 ctors
 
 
 -- TODO: Verify that Constructors aren't already in the global space or else throw a name clash error
 -- Use lookupADT for that
 resolveADTConstructor
-  :: FilePath -> ADTs -> Name -> [Name] -> Constructor -> Infer Vars
-resolveADTConstructor astPath tadts n params (Constructor cname cparams) = do
-  let t = buildADTConstructorReturnType astPath n params
-  t' <- mapM (argToType tadts n params) cparams
-  let ctype = foldr1 TArr (t' <> [t])
-  return $ M.fromList [(cname, Forall (TV <$> params) ctype)]
+  :: Env -> FilePath -> ADTs -> Name -> [Name] -> Constructor -> Infer Vars
+resolveADTConstructor priorEnv astPath tadts n params (Constructor cname cparams)
+  = do
+    let t = buildADTConstructorReturnType astPath n params
+    t' <- mapM (argToType priorEnv tadts n params) cparams
+    let ctype = foldr1 TArr (t' <> [t])
+    return $ M.fromList [(cname, Forall (TV <$> params) ctype)]
 
 buildADTConstructorReturnType :: FilePath -> Name -> [Name] -> Type
 buildADTConstructorReturnType astPath tname tparams =
   TComp astPath tname $ TVar . TV <$> tparams
 
 
 -- TODO: This should probably be merged with typingToType somehow
-argToType :: ADTs -> Name -> [Name] -> Typing -> Infer Type
-argToType tadts _ params (Meta _ _ (TRSingle n))
+argToType :: Env -> ADTs -> Name -> [Name] -> Typing -> Infer Type
+argToType _ tadts _ params (Meta _ _ (TRSingle n))
   | n == "Number" = return $ TCon CNum
   | n == "Boolean" = return $ TCon CBool
   | n == "String" = return $ TCon CString
   | isLower (head n) && (n `elem` params) = return $ TVar $ TV n
-  | isLower (head n) = throwError $ InferError (UnboundVariable n) NoReason
-  | otherwise = case M.lookup n tadts of
+  | isLower (head n) = newTVar
+  | -- A free var that is not in type params
+    otherwise = case M.lookup n tadts of
```

Comment:
possibly all of this is an error as a TRSingle can't be in tadts anyways ? It would then be a `TRComp []` ?